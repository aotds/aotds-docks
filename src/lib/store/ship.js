import { browser, dev } from "$app/environment";
import { readable, get, derived } from "svelte/store";
import { compose, applyMiddleware } from "redux";

import shipDux from "../shipDux/index";
import { initial } from "lodash";

let composeEnhancers = compose;

if (dev && browser && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) {
    composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;
}

export default (initialState = undefined) => {
    if (browser) {
        const i = localStorage.getItem("ship");

        if (i) initialState = JSON.parse(localStorage.getItem("ship"));
    }

    const duxStore = shipDux.createStore(initialState, (mw) =>
        composeEnhancers(applyMiddleware(mw))
    );

    let previous;
    const state = readable(duxStore.getState(), (set) => {
        duxStore.subscribe(() => {
            if (previous === duxStore.getState()) return;
            previous = duxStore.getState();
            set(previous);
            if (browser) localStorage.setItem("ship", JSON.stringify(previous));
        });
    });

    return {
        dispatch: duxStore.dispatch,
        state,
        shipMass: derived(state, (state) => state.reqs.mass),
    };
};
