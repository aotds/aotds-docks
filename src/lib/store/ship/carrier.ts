import Updux, { createPayloadAction } from "updux";
import u from "@yanick/updeep-remeda";
import { reqs, type Reqs } from "$lib/shipDux/reqs";

type Squadron = {
    type: string;
    reqs: Reqs;
};

const initialState = {
    nbrBays: 0,
    squadrons: [] as Squadron[],
    reqs,
};

export const squadronTypes = [
    { type: "standard", cost: 3 },
    { type: "fast", cost: 4 },
    { type: "heavy", cost: 5 },
    { type: "interceptor", cost: 3 },
    { type: "attack", cost: 4 },
    { type: "long range", cost: 4 },
    { type: "torpedo", cost: 6 },
];

const setNbrCarrierBays = createPayloadAction<number>("setNbrCarrierBays");
const setSquadronType = createPayloadAction(
    "setSquadronType",
    (id: number, type: string) => ({ id, type })
);

export const carrierDux = new Updux({
    initialState,
    actions: { setNbrCarrierBays, setSquadronType },
});

function calcBaysReqs(bays) {
    return {
        mass: 9 * bays,
        cost: 18 * bays,
    };
}

const adjustSquadrons = (bays: number) => (squadrons) => {
    if (squadrons.length === bays) return squadrons;

    if (squadrons.length > bays) {
        return squadrons.slice(0, bays);
    }

    return [
        ...squadrons,
        ...Array.from({ length: bays - squadrons.length })
            .fill({
                type: squadronTypes[0].type,
                reqs: {
                    cost: 6 * squadronTypes[0].cost,
                    mass: 6,
                },
            })
            .map((s, i) => ({ ...s, id: squadrons.length + i + 1 })),
    ];
};

carrierDux.addMutation(setNbrCarrierBays, (nbrBays) =>
    u({
        nbrBays,
        reqs: calcBaysReqs(nbrBays),
        squadrons: adjustSquadrons(nbrBays),
    })
);

carrierDux.addMutation(setSquadronType, ({ id, type }) => {
    return u({
        squadrons: u.map(
            u.if(u.matches({ id }), (state) => {
                return u(state, {
                    type,
                    reqs: squadronReqs(type),
                });
            })
        ),
    });
});

function squadronReqs(type: string) {
    return {
        mass: 6,
        cost: 6 * squadronTypes.find((s) => s.type === type)?.cost,
    };
}

/*
export const { actions, reducer } = createSlice({
    name: "carrier",
    initialStateState,
    reducers: {
        setCarrierBays: (state, action: PayloadAction<number>) => {
            state.bays = action.payload;
            state.reqs = calcBaysReqs(action.payload);
            state.squadrons = adjustSquadrons(action.payload)(state.squadrons);
        },
        setSquadronType: (
            state,
            action: PayloadAction<{ type: string; id: number }>
        ) => {
            state.squadrons[action.payload.id - 1] = {
                type: action.payload.type,
                reqs: squadronReqs(action.payload.type),
            };
        },
    },
});




*/
