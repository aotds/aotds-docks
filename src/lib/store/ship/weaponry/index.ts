import Updux, { createPayloadAction } from "updux";
import { adfcDux } from "./adfc";
import { fireconsDux } from "./firecons";
import { weaponsDux } from "./weapons";
import type { Reqs } from "$lib/shipDux/reqs";
import u from "@yanick/updeep-remeda";
import * as R from "remeda";

if (typeof process !== "undefined") {
    process.env.UPDEEP_MODE = "dangerously_never_freeze";
}

type MissileMagazine = {
    id: number;
    maxAmmo: number;
    extended: boolean;
};

const setMissileMagazine = createPayloadAction(
    "setMissileMagazine",
    (id, maxAmmo, extended = undefined) => ({ id, maxAmmo, extended })
);

const moveMissileMagazine = createPayloadAction(
    "moveMissileMagazine",
    (from, to) => ({
        from,
        to,
    })
);

const removeMissileMagazine = createPayloadAction<number>(
    "removeMissileMagazine"
);

const magazinesDux = new Updux({
    actions: { setMissileMagazine, removeMissileMagazine, moveMissileMagazine },
    initialState: [] as MissileMagazine[],
});

magazinesDux.addMutation(weaponsDux.actions.addWeapon, (payload) => (state) => {
    if (payload !== "sml") return state;

    state.push({
        id: state.length + 1,
        extended: false,
        maxAmmo: 1,
        reqs: magazineReqs({ extended: false, maxAmmo: 1 }),
    });
});

magazinesDux.addMutation(setMissileMagazine, (payload) =>
    u.map(
        u.if(u.matches({ id: payload.id }), (state) => {
            state = u(state, payload);
            return u(state, { reqs: magazineReqs(state) });
        })
    )
);
magazinesDux.addMutation(moveMissileMagazine, ({ from, to }) =>
    u.map(u.if(u.matches({ id: from }), { id: to }))
);

magazinesDux.addMutation(removeMissileMagazine, (id) =>
    u.reject(u.matches({ id }))
);

magazinesDux.addReaction((api) => (store) => {
    store
        .map(R.prop("id"))
        .filter(({ id }, i) => id !== i + 1)
        .forEach((id, i) => api.dispatch.moveMissileMagazine(id, i + 1));
});

function magazineReqs(magazine: MissileMagazine): Reqs {
    let mass = magazine.maxAmmo * (magazine.extended ? 3 : 2);
    return { mass, cost: 3 * mass };
}

export const weaponryDux = new Updux({
    initialState: {},
    subduxes: {
        adfc: adfcDux,
        firecons: fireconsDux,
        weapons: weaponsDux,
        missileMagazines: magazinesDux,
    },
});

weaponryDux.addMutation(moveMissileMagazine, ({ from, to }) =>
    u({
        weapons: u.map(
            u.if(u.matches({ specs: { missileMagazineId: from } }), {
                specs: { missileMagazineId: to },
            })
        ),
    })
);

weaponryDux.addReaction((api) => (state) => {
    const smls = state.weapons.filter(u.matches({ specs: { type: "sml" } }));

    const usedMagazines = smls.map(
        ({ specs: { missileMagazineId } }) => missileMagazineId
    );

    const unusedMags = state.missileMagazines
        .map(R.prop("id"))
        .filter((id) => !usedMagazines.includes(id));

    unusedMags.forEach((id) => api.dispatch.setMissileMagazine(id, 0));

    if (smls.length >= state.missileMagazines.length) return;

    api.dispatch.removeMissileMagazine(unusedMags[0]);
});
