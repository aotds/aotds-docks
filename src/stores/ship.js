import { browser } from "$app/env";
import { readable } from "svelte/store";
import { compose, applyMiddleware } from "redux";

import { calc_ship_req } from "../dux/utils";

let composeEnhancers = compose;

if (browser) {
  composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;
}

import shipDux from "../dux";

export default () => {
  let saved;

  if (browser) saved = window.localStorage.getItem("aotds-shipyard");

  if (saved) {
    saved = JSON.parse(saved);
  } else {
    saved = undefined;
  }

  const duxStore = shipDux.createStore(saved, (mw) =>
    composeEnhancers(applyMiddleware(mw))
  );

  duxStore.dispatch(
    duxStore.actions.set_ship_reqs(calc_ship_req(duxStore.getState()))
  );

  Object.entries(duxStore.actions).forEach(([type, action]) => {
    duxStore.dispatch[type] = (payload) => duxStore.dispatch(action(payload));
  });

  let previous = undefined;
  duxStore.subscribe(() => {
    let current = duxStore.getState();
    if (previous === current) return;
    previous = current;

    if (browser)
      window.localStorage.setItem("aotds-shipyard", JSON.stringify(current));
  });

  const state = readable(duxStore.getState(), (set) =>
    duxStore.subscribe(() => {
      set(duxStore.getState());
    })
  );

  const dispatch_action = (action, payload) =>
    duxStore.dispatch[action](payload);

  return {
    subscribe: state.subscribe,
    dispatch: duxStore.dispatch,
    actions: duxStore.actions,
    selectors: duxStore.selectors,
    dispatch_action,
  };
};
