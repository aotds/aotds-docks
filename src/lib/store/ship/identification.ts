import Updux, { createAction, withPayload } from "updux";
import u from "@yanick/updeep-remeda";
import { carrierDux } from "./carrier";

const initialState = {
  shipType: "",
  shipClass: "",
  isCarrier: false,
  reqs: {
    mass: 10,
    cost: 0,
    usedMass: 0,
  },
};

const setShipClass = createAction("setShipClass", withPayload<string>());
const updateIdentification = createAction("updateIdentification");
const setShipReqs = createAction("setShipReqs", withPayload());
const setCarrier = createAction("setCarrier", withPayload<boolean>());

export const identificationDux = new Updux({
  initialState,
  actions: {
    setShipClass,
    updateIdentification,
    setShipReqs,
    setCarrier,
  },
  selectors: {
    getShipMass: (state) => state.reqs.mass,
    isCarrier: ({ isCarrier }) => isCarrier,
  },
});

identificationDux.addMutation(setShipClass, (shipClass) => u({ shipClass }));
identificationDux.addMutation(updateIdentification, (update) => u(update));

identificationDux.addMutation(setCarrier, (isCarrier) => u({ isCarrier }));
identificationDux.addEffect(setCarrier, (api) => (next) => (action) => {
  next(action);
  if (!action.payload) {
    api.dispatch(carrierDux.actions.setNbrCarrierBays(0));
  }
});

identificationDux.addMutation(setShipReqs, (reqs) => u({ reqs }));

export default identificationDux;
