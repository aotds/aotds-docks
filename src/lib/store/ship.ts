import { createSelector } from "@reduxjs/toolkit";
import Updux from "updux";
import * as R from "remeda";

import identification from "./ship/identification";
import ftl, { calcFtlReqs } from "./ship/propulsion/ftl";
import drive from "./ship/propulsion/drive";
import { calcDriveReqs } from "$lib/shipDux/engine";
import { carrierDux } from "./ship/carrier";
import { streamliningDux as streamlining } from "./ship/structure/streamlining";
import { calcStreamliningReqs } from "./ship/structure/rules";

const shipDux = new Updux({
  subduxes: {
    identification,
    structure: new Updux({
      initialState: {},
      subduxes: {
        streamlining,
      },
    }),
    propulsion: new Updux({
      initialState: {},
      subduxes: {
        ftl,
        drive,
      },
    }),
    carrier: carrierDux,
  },
});

shipDux.addReaction((api) =>
  createSelector(
    api.selectors.getFtlType,
    api.selectors.getShipMass,
    (type, mass) => api.dispatch.setFtlReqs(calcFtlReqs(type, mass))
  )
);

shipDux.addReaction((api) => (state) => {
  let cost = 0;
  let mass = 0;

  let subsystems = R.values(R.omit(state, ["identification"]));

  while (subsystems.length > 0) {
    const subsystem = subsystems.shift();
    if (typeof subsystem !== "object") continue;

    if (subsystem.reqs) {
      cost += subsystem.reqs.cost;
      mass += subsystem.reqs.mass;
    }

    subsystems.push(...Object.values(subsystem));
  }

  api.dispatch.setShipReqs({ cost, usedMass: mass });
});

shipDux.addReaction((api) =>
  createSelector(
    api.selectors.getShipMass,
    (state) => state.propulsion.drive.rating,
    (state) => state.propulsion.drive.advanced,
    (mass, rating, advanced) =>
      api.dispatch.setDriveReqs(calcDriveReqs(mass, rating, advanced))
  )
);

shipDux.addReaction((api) =>
  createSelector(
    //    (state) => state,
    api.selectors.getShipMass,
    api.selectors.getStreamlining,
    (mass, type) => {
      console.log("AH!", mass, type);
      api.dispatch.setStreamliningReqs(calcStreamliningReqs(type, mass));
    }
  )
);

export default shipDux;
