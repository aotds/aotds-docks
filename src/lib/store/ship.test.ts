import { test, expect } from "vitest";
import ship from "./ship";

test("misc", () => {
    const store = ship.createStore();
    store.dispatch.setFtlType("standard");

    expect(store.getState().propulsion.ftl.reqs.mass).toEqual(1);
    expect(store.getState().identification.reqs.usedMass).toEqual(1);

    store.dispatch.setDrive({ rating: 3, advanced: true });

    expect(store.getState().propulsion.drive.reqs).toEqual({ mass: 2, cost: 6 });
});
