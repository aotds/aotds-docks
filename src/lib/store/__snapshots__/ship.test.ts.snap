// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`state has the expected shape 1`] = `
{
  "carrier": {
    "nbrBays": 0,
    "reqs": {
      "cost": 0,
      "mass": 0,
    },
    "squadrons": [],
  },
  "identification": {
    "isCarrier": false,
    "reqs": {
      "cost": 0,
      "mass": 10,
      "usedMass": 0,
    },
    "shipClass": "",
    "shipType": "",
  },
  "propulsion": {
    "drive": {
      "advanced": false,
      "rating": 0,
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
    },
    "ftl": {
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
      "type": "none",
    },
  },
  "schemaVersion": "1",
  "structure": {
    "armor": {
      "layers": [],
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
    },
    "cargo": {
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
      "space": 0,
    },
    "hull": {
      "max": 0,
      "min": 0,
      "rating": 0,
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
    },
    "screens": {
      "advanced": 0,
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
      "standard": 0,
    },
    "streamlining": {
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
      "type": "none",
    },
  },
  "weaponry": {
    "adfc": {
      "rating": 0,
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
    },
    "firecons": {
      "reqs": {
        "cost": 0,
        "mass": 0,
      },
      "stations": 0,
    },
    "weapons": [],
  },
}
`;
